src/charm.py

```diff
@@ -255,6 +255,102 @@ def _loki_config(self) -> dict:
         )
         return yaml.safe_load(config)
 
+    def _loki_push_api_alert_rules_changed(self, _: LokiPushApiAlertRulesChanged) -> None:
+        """Perform all operations needed to keep alert rules in the right status."""
+        if self._ensure_alert_rules_path():
+            self._regenerate_alert_rules()
+
+            # Don't try to configure if checking the rules left us in BlockedStatus
+            if isinstance(self.unit.status, ActiveStatus):
+                self._configure()
+
+    def _ensure_alert_rules_path(self) -> bool:
+        """Ensure that the workload container has the appropriate directory structure."""
+        # create tenant dir so that the /loki/api/v1/rules endpoint returns "no rule groups found"
+        # instead of "unable to read rule dir /loki/rules/fake: no such file or directory"
+        if self._container.can_connect():
+            try:
+                self._container.make_dir(self._rules_dir, make_parents=True)
+                return True
+            except (FileNotFoundError, ProtocolError, PathError):
+                logger.debug("Could not create loki directory.")
+                return False
+            except Exception as e:
+                logger.debug("Could create loki directory: %s", e)
+                return False
+        return False
+
+    def _regenerate_alert_rules(self):
+        """Recreate all alert rules."""
+        self._remove_alert_rules_files()
+        # If there aren't any alerts, we can just clean it and move on
+        if self.loki_provider.alerts:
+            self._generate_alert_rules_files()
+            self._check_alert_rules()
+
+    def _generate_alert_rules_files(self) -> None:
+        """Generate and upload alert rules files.
+
+        Args:
+            container: Container into which alert rules files are going to be uploaded
+        """
+        file_mappings = {}
+
+        for identifier, alert_rules in self.loki_provider.alerts.items():
+            rules = yaml.dump({"groups": alert_rules["groups"]})
+            file_mappings["{}_alert.rules".format(identifier)] = rules
+
+        if self._container.can_connect():
+            for filename, content in file_mappings.items():
+                path = os.path.join(self._rules_dir, filename)
+                self._container.push(path, content, make_dirs=True)
+        logger.debug("Saved alert rules to disk")
```

Comment:
We could make this more useful:
```suggestion
        logger.debug("Saved %s alert rules to disk: %s", len(file_mappings), self._rules_dir)
```